include/mbgl/gfx/vertex_attribute.hpp

```diff
@@ -0,0 +1,359 @@
+                                                int index = -1,
+                                                AttributeDataType = AttributeDataType::Invalid,
+                                                std::size_t count = 1);
+
+    /// Add a new attribute element if it doesn't already exist.
+    /// Returns a pointer to the new element on success, or null if the type or count conflict with an existing entry.
+    /// The result is valid only until the next non-const method call on this class.
+    /// @param index index to match, or -1 for any
+    /// @param type type to match, or `Invalid` for any
+    /// @param count type to match, or 0 for any
+    const std::unique_ptr<VertexAttribute>& getOrAdd(std::string name,
+                                                     int index = -1,
+                                                     AttributeDataType type = AttributeDataType::Invalid,
+                                                     std::size_t count = 0);
+
+    // Set a value if the element is present
+    template <typename T>
+    bool set(const std::string& name, std::size_t i, T value) {
+        if (const auto& item = get(name)) {
+            return item->set(i, value);
+        }
+        return false;
+    }
+
+    /// Indicates whether any values have changed
+    bool isDirty() const {
+        return std::any_of(
+            attrs.begin(), attrs.end(), [](const auto& kv) { return kv.second && kv.second->isDirty(); });
+    }
+
+    void clear();
+
+    /// Do something with each attribute
+    void observeAttributes(const std::function<void(const std::string&, VertexAttribute&)>& f) {
+        std::for_each(attrs.begin(), attrs.end(), [&](const auto& kv) {
+            if (kv.second) {
+                f(kv.first, *kv.second);
+            }
+        });
+    }
+    void observeAttributes(const std::function<void(const std::string&, const VertexAttribute&)>& f) const {
+        std::for_each(attrs.begin(), attrs.end(), [&](const auto& kv) {
+            if (kv.second) {
+                f(kv.first, *kv.second);
+            }
+        });
+    }
+
+    using ResolveDelegate =
+        std::function<void(const std::string&, VertexAttribute&, const std::unique_ptr<VertexAttribute>&)>;
+    /// Call the provided delegate with each value, providing the override if one exists.
+    void resolve(const VertexAttributeArray& overrides, ResolveDelegate) const;
+
+    VertexAttributeArray& operator=(VertexAttributeArray&&);
+    VertexAttributeArray& operator=(const VertexAttributeArray&);
+
+    virtual std::unique_ptr<VertexAttributeArray> clone() const {
+        auto newAttrs = std::make_unique<VertexAttributeArray>();
+        newAttrs->copy(*this);
+        return newAttrs;
+    }
+
+    template <class... DataDrivenPaintProperty, class Binders, class Evaluated>
+    std::vector<std::string> readDataDrivenPaintProperties(const Binders& binders, const Evaluated& evaluated) {
```

Comment:
This one definitely needs a documentation comment.

Some comments to help a future programmer make sense of the the internals also would not hurt!

Reply:
I've also wondered about the comments in the code. What is the policy in this repo about comments?
I see that in some files nearly everything has doxygen comments. In some files only some minor things have comments.
Also in different files different style of doxygen comments is used.

```
class RenderTarget {
public:
    virtual ~RenderTarget() = default;

    void setTexture(const gfx::Texture2DPtr& texture_) { texture = std::move(texture_); };
    const gfx::Texture2DPtr& getTexture() const { return texture; };

    bool addLayerGroup(LayerGroupBasePtr, bool replace);
    bool removeLayerGroup(const int32_t layerIndex);
    size_t numLayerGroups() const noexcept;
    const LayerGroupBasePtr& getLayerGroup(const int32_t layerIndex) const;
    void observeLayerGroups(std::function<void(LayerGroupBase&)>);
    void observeLayerGroups(std::function<void(const LayerGroupBase&)>) const;

    virtual void upload(gfx::UploadPass& uploadPass) = 0;
    virtual void render(RenderOrchestrator&, const RenderTree&, PaintParameters&) = 0;

protected:
    gfx::Texture2DPtr texture;
    using LayerGroupMap = std::map<int32_t, LayerGroupBasePtr>;
    LayerGroupMap layerGroupsByLayerIndex;
};
```

then this

```
/// The RenderFrontend is the bridge between the Map and
/// platform used to update and observer the Renderer
///
/// It hides any threading specifics and always replies on
/// the original thread.
class RendererFrontend {
public:
    virtual ~RendererFrontend() = default;

    /// Must synchronously clean up the Renderer if set
    virtual void reset() = 0;

    /// Implementer must bind the renderer observer to the renderer in a
    /// appropriate manner so that the callbacks occur on the main thread
    virtual void setObserver(RendererObserver&) = 0;

    /// Coalescing updates is up to the implementer
    virtual void update(std::shared_ptr<UpdateParameters>) = 0;
};
```

then this

```
    // Debug
    void dumpDebugLogs();

    /**
     * @brief In Tile map mode, enables or disables collecting of the placed
     * symbols data, which can be obtained with `getPlacedSymbolsData()`.
     *
     * The placed symbols data collecting is disabled by default.
     */
    void collectPlacedSymbolData(bool enable);

    /**
     * @brief If collecting of the placed symbols data is enabled, returns the
     * reference to the `PlacedSymbolData` vector holding the collected data.
     *
     * Note: the returned vector gets re-populated at every `render()` call.
     *
     * @return collected placed symbols data
     */
    const std::vector<PlacedSymbolData>& getPlacedSymbolsData() const;
```

Are there some strict rules about doxygen comments or they are relaxed (if doxygen works without errors then it's okay). And also what are the rules (if there are any) about what should be commented/explained and what should not?

Reply:
@VladyslavOdobesku-TomTom Any non-trivial methods and classes part of the public API should have some documentation comments.

I think the general pattern is that triple slash comments is used when something only has a description whereas `/**` is used when other things are included such as documentation for parameters or the return value.